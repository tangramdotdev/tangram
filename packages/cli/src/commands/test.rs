use crate::Cli;
use tangram_client as tg;
use tangram_error::Result;

/// Build the target named "test" from the specified package.
#[derive(Debug, clap::Args)]
#[command(verbatim_doc_comment)]
pub struct Args {
	/// If this flag is set, then the command will exit immediately instead of waiting for the build's output.
	#[arg(short, long)]
	pub detach: bool,

	/// The path to the executable in the artifact to run.
	#[arg(long)]
	pub executable_path: Option<tg::Path>,

	/// If this flag is set, the package's lockfile will not be updated.
	#[arg(long)]
	pub locked: bool,

	/// The package to build.
	#[arg(short, long, default_value = ".")]
	pub package: tg::Dependency,

	/// The retry strategy to use.
	#[arg(long, default_value_t)]
	pub retry: tg::build::Retry,
}

impl Cli {
	pub async fn command_test(&self, args: Args) -> Result<()> {
		// Create the build args.
		let args = super::build::Args {
			detach: args.detach,
			locked: args.locked,
			no_tui: false,
			output: None,
			package: args.package,
			retry: args.retry,
			target: "test".to_owned(),
		};

		// Build!
		self.command_build(args).await?;

		Ok(())
	}
}
