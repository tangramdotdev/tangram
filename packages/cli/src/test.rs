use crate::Config;
use futures::{Future, FutureExt as _};
use std::{panic::AssertUnwindSafe, sync::Arc};
use tangram_client as tg;
use tangram_temp::Temp;
use url::Url;

pub async fn test<F, Fut>(tg: &'static str, f: F)
where
	F: FnOnce(Arc<tokio::sync::Mutex<Context>>) -> Fut,
	Fut: Future<Output = ()> + Send,
{
	// Create the context.
	let context = Arc::new(tokio::sync::Mutex::new(Context::new(tg)));

	// Run the test and catch a panic if one occurs.
	let result = AssertUnwindSafe(f(context.clone())).catch_unwind().await;

	// Handle the result.
	let context = context.lock().await;
	match result {
		Ok(()) => {
			// If there was no panic, then gracefully shutdown the servers.
			for server in &context.servers {
				let mut process = server.process.lock().await;
				unsafe { libc::kill(process.id().unwrap().try_into().unwrap(), libc::SIGINT) };
				let result = process.wait().await;
				if let Ok(status) = result {
					if !status.success() {}
				}
			}
		},
		Err(error) => {
			// If there was a panic, then forcefully shutdown the servers and resume unwinding.
			for server in &context.servers {
				let mut process = server.process.lock().await;
				unsafe { libc::kill(process.id().unwrap().try_into().unwrap(), libc::SIGKILL) };
				process.wait().await.ok();
			}
			std::panic::resume_unwind(error);
		},
	}
}

pub struct Context {
	servers: Vec<Arc<Server>>,
	tg: &'static str,
}

pub struct Server {
	config: Config,
	process: tokio::sync::Mutex<tokio::process::Child>,
	temp: Arc<Temp>,
	tg: &'static str,
}

impl Context {
	#[must_use]
	pub fn new(tg: &'static str) -> Self {
		Self {
			servers: Vec::new(),
			tg,
		}
	}

	pub async fn spawn_server(&mut self) -> tg::Result<Arc<Server>> {
		let server = Server::new(self.tg).await?;
		let server = Arc::new(server);
		self.servers.push(server.clone());
		Ok(server)
	}

	pub async fn spawn_server_with_config(&mut self, config: Config) -> tg::Result<Arc<Server>> {
		let server = Server::with_config(self.tg, config).await?;
		let server = Arc::new(server);
		self.servers.push(server.clone());
		Ok(server)
	}

	pub async fn spawn_server_with_temp_and_config(
		&mut self,
		config: Config,
	) -> tg::Result<Arc<Server>> {
		let server = Server::with_config(self.tg, config).await?;
		let server = Arc::new(server);
		self.servers.push(server.clone());
		Ok(server)
	}
}

impl Server {
	async fn new(tg: &'static str) -> tg::Result<Self> {
		let config = Config::default();
		Self::with_config(tg, config).await
	}

	async fn with_config(tg: &'static str, config: Config) -> tg::Result<Self> {
		// Create a temp and create the config and data paths.
		let temp = Arc::new(Temp::new());
		tokio::fs::create_dir_all(temp.path())
			.await
			.map_err(|source| tg::error!(!source, "failed to create the directory"))?;
		let config_path = temp.path().join(".config/tangram/config.json");
		tokio::fs::create_dir_all(config_path.parent().unwrap())
			.await
			.map_err(|source| tg::error!(!source, "failed to create the config directory"))?;
		let config_json = serde_json::to_vec_pretty(&config).unwrap();
		tokio::fs::write(&config_path, config_json)
			.await
			.map_err(|source| tg::error!(!source, "failed to write the config"))?;
		let data_path = temp.path().join(".tangram");
		tokio::fs::create_dir_all(&data_path)
			.await
			.map_err(|source| tg::error!(!source, "failed to create the data directory"))?;

		// Create the command.
		let log_path = temp.path().join(".tangram/log");
		let stdout = tokio::fs::OpenOptions::new()
			.create(true)
			.write(true)
			.truncate(true)
			.open(&log_path)
			.await
			.map_err(|source| tg::error!(!source, "failed to open the log file"))?
			.into_std()
			.await;
		let stderr = tokio::fs::OpenOptions::new()
			.create(true)
			.write(true)
			.truncate(true)
			.open(&log_path)
			.await
			.map_err(|source| tg::error!(!source, "failed to open the log file"))?
			.into_std()
			.await;
		let mut command = tokio::process::Command::new(tg);
		command.stdout(stdout);
		command.stderr(stderr);
		command.arg("--config");
		command.arg(&config_path);
		command.arg("--path");
		command.arg(data_path);
		command.arg("serve");

		// Spawn the process.
		let process = command.spawn().unwrap();
		let process = tokio::sync::Mutex::new(process);

		// Create the server.
		let server = Self {
			config,
			process,
			temp,
			tg,
		};

		Ok(server)
	}

	#[must_use]
	pub fn tg(&self) -> tokio::process::Command {
		let mut command = tokio::process::Command::new(self.tg);
		let config_path = self.temp.path().join(".config/tangram/config.json");
		let data_path = self.temp.path().join(".tangram");
		command
			.arg("--config")
			.arg(config_path)
			.arg("--path")
			.arg(data_path)
			.arg("--mode")
			.arg("client");
		command
	}

	#[must_use]
	pub fn config(&self) -> &Config {
		&self.config
	}

	#[must_use]
	pub fn process(&self) -> &tokio::sync::Mutex<tokio::process::Child> {
		&self.process
	}

	#[must_use]
	pub fn url(&self) -> Url {
		let path = self.temp.path().join(".tangram/socket");
		let path = path.to_str().unwrap();
		let path = urlencoding::encode(path);
		format!("http+unix://{path}").parse().unwrap()
	}

	pub async fn print_log(&self) -> tg::Result<()> {
		let log_path = self.temp.join(".tangram/log");
		let mut log_file = tokio::fs::File::open(log_path)
			.await
			.map_err(|source| tg::error!(!source, "failed to open the log file"))?;
		let mut stderr = tokio::io::stderr();
		tokio::io::copy(&mut log_file, &mut stderr).await.ok();
		Ok(())
	}
}

#[macro_export]
macro_rules! assert_output_success {
	($output:expr) => {
		let output = &$output;
		if !output.status.success() {
			let mut stderr = tokio::io::stderr();
			stderr.write_all(&output.stderr).await.unwrap();
			stderr.flush().await.unwrap();
		}
		assert!(output.status.success());
	};
}
pub use assert_output_success;
