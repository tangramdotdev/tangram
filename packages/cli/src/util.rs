use console::style;
use either::Either;
use tangram_client as tg;
use tangram_error::Error;

pub fn build_or_object_id(s: &str) -> Result<Either<tg::build::Id, tg::object::Id>, String> {
	if let Ok(value) = s.parse() {
		return Ok(Either::Left(value));
	}
	if let Ok(value) = s.parse() {
		return Ok(Either::Right(value));
	}
	Err("Failed to parse.".to_string())
}

pub fn print_error(mut error: &tangram_error::Error) {
	eprintln!("{}:", style("Error").red());
	let mut first = true;
	loop {
		if !first {
			eprintln!();
		}
		first = false;
		eprint!("{}", style("->").red());
		let Error {
			message,
			location,
			stack,
			source,
			values,
		} = error;
		eprint!(" {message}");
		if let Some(location) = &location {
			let location = style(location).dim().white();
			eprint!(" {location}");
		}
		for (name, value) in values {
			eprintln!();
			let name = style(name).blue();
			let value = style(value).green();
			eprint!("   {name} = {value}");
		}
		for location in stack.iter().flatten() {
			eprintln!();
			let location = style(location).dim().white();
			eprint!(" {location}");
		}
		if let Some(source) = &source {
			error = source;
		} else {
			break;
		}
	}
	eprintln!();
}
