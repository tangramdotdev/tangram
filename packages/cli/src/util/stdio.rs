use bytes::Bytes;
use futures::{Stream, TryStreamExt as _};
use std::io::Read as _;
use tangram_client as tg;
use tokio::io::AsyncBufRead;
use tokio_stream::wrappers::ReceiverStream;
use tokio_util::io::StreamReader;

pub(crate) fn stdin() -> impl AsyncBufRead + Send + 'static {
	StreamReader::new(stdin_stream().map_err(std::io::Error::other))
}

pub(crate) fn stdin_stream() -> impl Stream<Item = tg::Result<Bytes>> + Send + 'static {
	let (send, recv) = tokio::sync::mpsc::channel(1);
	std::thread::spawn(move || {
		let mut stdin = std::io::stdin();
		loop {
			let mut buf = vec![0u8; 4096];
			let result = match stdin.read(&mut buf) {
				Ok(0) => break,
				Ok(n) => Ok(Bytes::copy_from_slice(&buf[0..n])),
				Err(source) => Err(tg::error!(!source, "failed to read stdin")),
			};
			let result = send.blocking_send(result);
			if result.is_err() {
				break;
			}
		}
	});
	ReceiverStream::new(recv)
}
