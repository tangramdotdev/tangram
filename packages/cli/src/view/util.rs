use ratatui::{prelude::*, widgets::Block};

pub fn render_block_and_get_area(title: &str, focused: bool, area: Rect, buf: &mut Buffer) -> Rect {
	let block = Block::bordered()
		.title(title)
		.border_style(Style::default().fg(if focused { Color::Blue } else { Color::White }));
	block.render(area, buf);
	Layout::default()
		.constraints([Constraint::Percentage(100)])
		.margin(1)
		.split(area)
		.first()
		.copied()
		.unwrap_or(area)
}
