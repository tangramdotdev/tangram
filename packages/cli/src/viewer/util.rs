use {
	futures::TryStreamExt as _,
	std::{io::Write as _, pin::pin},
	tangram_client::{self as tg, handle::Ext as _},
};

const BLOB_LENGTH_LIMIT: u64 = 1 << 20;

pub async fn format_blob<H: tg::Handle>(handle: &H, blob: &tg::Blob) -> tg::Result<String> {
	let length = blob.length(handle).await?;
	if length > BLOB_LENGTH_LIMIT {
		return Err(tg::error!("cannot view blobs larger than 1Mib"));
	}
	let mut contents = Vec::new();
	let arg = tg::read::Arg {
		blob: blob.id(),
		options: tg::read::Options {
			length: Some(BLOB_LENGTH_LIMIT),
			..tg::read::Options::default()
		},
	};
	let stream = handle
		.try_read(arg)
		.await?
		.ok_or_else(|| tg::error!("failed to read the blob"))?;
	let mut stream = pin!(stream);
	let mut is_utf8 = true;
	while let Some(chunk) = stream.try_next().await? {
		if let (true, Ok(chunk)) = (is_utf8, std::str::from_utf8(&chunk.bytes)) {
			contents.extend_from_slice(chunk.as_bytes());
			continue;
		}
		is_utf8 = false;
		contents.reserve(chunk.bytes.len());
		for byte in chunk.bytes {
			if byte.is_ascii_graphic() {
				contents.push(byte);
			} else {
				write!(&mut contents, " {byte:20x}").unwrap();
			}
		}
	}
	Ok(String::from_utf8(contents).unwrap())
}
