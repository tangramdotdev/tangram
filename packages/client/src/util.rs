pub mod arc;
pub mod fs;
pub mod http;
pub mod serde;
