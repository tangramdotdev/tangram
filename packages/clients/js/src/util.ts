import type * as tg from "./index.ts";

export type MaybePromise<T> = T | Promise<T>;

export type MaybeMutation<T extends tg.Value = tg.Value> = T | tg.Mutation<T>;

export type MutationMap<
	T extends { [key: string]: tg.Value } = { [key: string]: tg.Value },
> = {
	[K in keyof T]?: tg.Mutation<T[K]>;
};

export type MaybeMutationMap<
	T extends { [key: string]: tg.Value } = { [key: string]: tg.Value },
> = {
	[K in keyof T]?: tg.MaybeMutation<T[K]>;
};

export type ValueOrMaybeMutationMap<T extends tg.Value = tg.Value> = T extends
	| undefined
	| boolean
	| number
	| string
	| tg.Object
	| Uint8Array
	| tg.Mutation
	| tg.Template
	| Array<infer _U extends tg.Value>
	? T
	: T extends { [key: string]: tg.Value }
		? {
				[K in keyof T]?: T[K] | tg.Mutation<T[K]>;
			}
		: never;

export type MaybeReferent<T> = T | tg.Referent<T>;

export type UnresolvedArgs<T extends Array<tg.Value>> = {
	[K in keyof T]: tg.Unresolved<T[K]>;
};

export type ResolvedArgs<T extends Array<tg.Unresolved<tg.Value>>> = {
	[K in keyof T]: tg.Resolved<T[K]>;
};

export type ReturnValue = tg.MaybePromise<void> | tg.Unresolved<tg.Value>;

export type ResolvedReturnValue<T extends tg.ReturnValue> =
	T extends tg.MaybePromise<void>
		? undefined
		: T extends tg.Unresolved<tg.Value>
			? tg.Resolved<T>
			: never;
