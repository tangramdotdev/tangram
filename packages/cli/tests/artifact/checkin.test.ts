import { expect, test } from "bun:test";
import Server from "../server.ts";
import { compare, directory, file } from "../util.ts";

test("file", async () => {
	await using server = await Server.start();
	let dir = await directory({
		"hello.txt": "Hello, World!",
	});
	let id = await server.tg`checkin ${dir}/hello.txt`
		.text()
		.then((t) => t.trim());
	expect(id).toMatchSnapshot();
});

test("executable", async () => {
	await using server = await Server.start();
	let dir = await directory({
		executable: file({ contents: "", executable: true }),
	});
	let id = await server.tg`checkin ${dir}/executable`
		.text()
		.then((t) => t.trim());
	expect(id).toMatchSnapshot();
});

test("roundtrip directory", async () => {
	await using server = await Server.start();
	let dir = await directory({
		"hello.txt": "Hello, World!",
	});
	let id = await server.tg`checkin ${dir}`.text().then((t) => t.trim());
	let path = await server.tg`checkout ${id}`.text().then((t) => t.trim());
	const equal = await compare(path, dir);
	expect(equal).toBeTrue();
});
