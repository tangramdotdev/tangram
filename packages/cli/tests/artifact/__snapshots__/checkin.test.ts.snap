// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`directory 1`] = `"dir_019400zf13yggz7dx563tajyet2vhxcj0rrgw2wx3yjrsvt51ygjb0"`;

exports[`directory 2`] = `"{"entries":{"child":"dir_01vm1rf4hw6yr8pvj9pct7cmvfejb9m1kh0p9te9fzxt1r132dpqn0","hello.txt":"fil_01my5jh8zn8r1jmpm8j6286w0v65pqe95wmn116pvf6k18rkbbacf0","link":"sym_012wvrd2ne7t45g7424e0fqy3wfkkb0wk3k220qavx9p68yqspr380"}}"`;

exports[`directory 3`] = `
"{
  "complete": true,
  "count": 11,
  "depth": 6,
  "weight": 620
}"
`;

exports[`file 1`] = `"fil_01tvcqmbbf8dkkejz6y69ywvgfsh9gyn1xjweyb9zgv0sf4752446g"`;

exports[`file 2`] = `"{"contents":"lef_015258d9wz42hxdq6ds9vh7fnet5w7k0mpqqx7j4zt59hdjwkvz3w0"}"`;

exports[`file 3`] = `
"{
  "complete": true,
  "count": 2,
  "depth": 2,
  "weight": 86
}"
`;

exports[`symlink 1`] = `"sym_01ejhx3jsq523gdc0cwg7m5ghe0df2n490jzbje33f99hgxby0y290"`;

exports[`symlink 2`] = `"{"artifact":"fil_019gsxqq56y3655ha3qt5rztxhgh3cxcwxts9sd6ge9t6qaevj92zg"}"`;

exports[`symlink 3`] = `
"{
  "complete": true,
  "count": 3,
  "depth": 3,
  "weight": 150
}"
`;

exports[`cycle 1`] = `"dir_01jtzwbst0gs09fy5ee8jaj2py247t7wmfjsaegfm9y1ggv49m21j0"`;

exports[`cycle 2`] = `"{"graph":"gph_01s75pva4t48fx6rjdjzae1tmezjfhfcg3s9k982hrjxevcnm4d5dg","node":1}"`;

exports[`cycle 3`] = `
"{
  "complete": true,
  "count": 2,
  "depth": 2,
  "weight": 162
}"
`;

exports[`cyclic-path-dependencies 1`] = `"dir_01rzyxgmtg8dq78f6kaafxn4dvvmege488yh8q4zyn8hgd0228xf20"`;

exports[`cyclic-path-dependencies 2`] = `"{"entries":{"dependency.tg.ts":"fil_01b3snp6y4yxpqzj373q2awafxqvvfy6k2n19f92d3ptn157dpxm70","tangram.ts":"fil_01ay4yr59c3bh5th3e5pdw243jjneywzyxx096qqzc1v6bmfs3he20"}}"`;

exports[`cyclic-path-dependencies 3`] = `
"{
  "complete": true,
  "count": 9,
  "depth": 3,
  "weight": 1629
}"
`;

exports[`executable 1`] = `"fil_017ywcxv0d0f712w8c7z77szz4emsr533se6jjv27xrcj670k30av0"`;

exports[`executable 2`] = `"{"contents":"lef_01nw9mkefnz6gtd8209qn3dq6996dwp9e9nq0h5dycka9wns0z69h0","executable":true}"`;

exports[`executable 3`] = `
"{
  "complete": true,
  "count": 2,
  "depth": 2,
  "weight": 91
}"
`;

exports[`depth 1`] = `"dir_01fkx1ydhprxsxq7peg3vzwasgb1z94hcmn5gjxqnn3dmm6vr9xt30"`;

exports[`depth 2`] = `"{"entries":{"nested":"dir_01t197hbevadkmr7bgvn0e633151e2rs4gecrk5010eae98ds5b5m0"}}"`;

exports[`depth 3`] = `
"{
  "complete": true,
  "count": 5,
  "depth": 5,
  "weight": 347
}"
`;
